src/pyoncatqt/login.py

```diff
@@ -0,0 +1,315 @@
+            return
+        except pyoncat.LoginRequiredError:
+            self.show_message("A username and/or password was not provided when logging in.")
+            self.user_pwd.setText("")
+            return
+
+        self.login_status.emit(True)
+        # close dialog
+        self.close()
+
+
+class ONCatLogin(QGroupBox):
+    """
+    ONCatLogin widget for connecting to the ONCat service.
+    This widget provides a label and a button to call the login dialog.
+
+    Params
+    ------
+    key : str, required
+        The key used to retrieve ONCat client ID from configuration. Defaults to None.
+    parent : QWidget, optional
+        The parent widget.
+    kwargs : Dict[str, Any], optional
+        Additional keyword arguments.
+
+    Attributes
+    ----------
+    connection_updated : Signal
+        Signal emitted when the connection status is updated.
+
+    Methods
+    -------
+    update_connection_status() -> None:
+        Update the connection status.
+    is_connected() -> bool:
+        Check if connected to OnCat.
+    get_agent_instance() -> pyoncat.ONCat:
+        Get the OnCat agent instance.
+    connect_to_oncat() -> None:
+        Connect to OnCat.
+    read_token() -> dict:
+        Read token from file.
+    write_token(token: dict) -> None:
+        Write token to file.
+    """
+
+    connection_updated = Signal(bool)
+
+    def __init__(self: QGroupBox, key: str = None, parent: QWidget = None, **kwargs: Dict[str, Any]) -> None:
+        """
+        Initialize the ONCatLogin widget.
+
+        Params
+        ------
+        key : str, optional
+            The key used to retrieve ONCat client ID from configuration. Defaults to None.
+        parent : QWidget, optional
+            The parent widget.
+        **kwargs : Dict[str, Any], optional
+            Additional keyword arguments.
+        """
+        super().__init__(parent)
+        self.oncat_options_layout = QGridLayout()
+        self.setLayout(self.oncat_options_layout)  # Set the layout for the group box
+
+        # Status indicator (disconnected: red, connected: green)
+        self.status_label = QLabel("")
+        self.status_label.setToolTip("ONCat connection status.")
+        self.oncat_options_layout.addWidget(self.status_label, 4, 0)
+
+        # Connect to OnCat button
+        self.oncat_button = QPushButton("&Connect to ONCat")
+        self.oncat_button.setFixedWidth(300)
+        self.oncat_button.setToolTip("Connect to ONCat (requires login credentials).")
+        self.oncat_button.clicked.connect(self.connect_to_oncat)
+        self.oncat_options_layout.addWidget(self.oncat_button, 4, 1)
+
+        self.error_message_callback = None
+
+        # OnCat agent
+
+        self.oncat_url = get_data("login.oncat", "oncat_url")
+        self.client_id = get_data("login.oncat", f"{key}_id")
+        if not self.client_id:
+            raise ValueError(f"Invalid key: {key}, ONCat client ID not found")
```

Comment:
Can you put a more descriptive message: e.g. Invalid module {key}. No OnCat client Id is found for this application.